### PR TITLE
Py bind iv

### PR DESCRIPTION
**Pull request by mjmottram**
Hopefully I retained all my changes from the PR against the pre-rebase version.  Overloaded functions now included.  The simple gaussian_smear.py script works for me, nothing else properly tested.

### CODE REVIEW

**Comment by mjmottram** [comment]
OK, I've tested this with the simple fit and gaussian smear and both seem to be working.  I think it's good to review; I could look at adding python unit tests at another point.

File: src/teststat/ChiSquare.cpp [hunk 1]
```diff
@@ -3,6 +3,12 @@
 #include <sstream>
 #include <Exceptions.h>
 
+ChiSquare::~ChiSquare(){
```
Comment (jackdunger) [inline src/teststat/ChiSquare.cpp]: Please remove this method, the test statistics do not own the data

File: src/teststat/BinnedNLLH.cpp [hunk 2]
```diff
@@ -116,6 +116,12 @@ BinnedNLLH::AddSystematic(Systematic* sys_){
 }
 
 void
+BinnedNLLH::SetDataSet(DataSet& dataSet_){
```
Comment (jackdunger) [inline src/teststat/BinnedNLLH.cpp]: this is duplicate of the method below

File: src/plot/ROOTMultiPlot.cpp [hunk 3]
```diff
@@ -77,7 +77,10 @@ ROOTMultiPlot::Construct(){
   for(size_t i = 0; i < fNames.size(); i++){
     fLegend.AddEntry(&fHists.at(i), fNames.at(i).c_str());
     fHists[i].SetLineColor(fColorScheme[i %fNcolors]);
-    fHists.at(i).Draw("same");
+    if(i!=0)
```
Comment (jackdunger) [inline src/plot/ROOTMultiPlot.cpp]: I'm pretty sure this is already the default behaviour, Draw("same") behaves like Draw() if nothing has been drawn before

File: src/teststat/ChiSquare.h [hunk 4]
```diff
@@ -15,6 +15,7 @@ class DataSet;
 class ChiSquare : public TestStatistic{
  public:
     ChiSquare() : fCalculatedDataPdf(false) {}
+    ~ChiSquare();
```
Comment (jackdunger) [inline src/teststat/ChiSquare.h]: the default destructor should be fine for this class, there's no memory management

**Comment by jackdunger** [comment]
@mjmottram thanks! I've commented on some changes you made to the C++. When I get the chance i'll try out building and running it

File: src/plot/ROOTMultiPlot.cpp (hunk 3 above)
Comment (mjmottram) [inline src/plot/ROOTMultiPlot.cpp]: Hmm, it created a blank canvas for me, let me check...

File: src/teststat/BinnedNLLH.cpp (hunk 2 above)
Comment (mjmottram) [inline src/teststat/BinnedNLLH.cpp]: boost::python can only handle pass by reference, it either needs a confusing wrapper function that I still haven't gotten to work within the binding implementation or a simple extra method like this in the class.

File: src/plot/ROOTMultiPlot.cpp (hunk 3 above)
Comment (mjmottram) [inline src/plot/ROOTMultiPlot.cpp]: OK, I'll revert it - I did get a blank canvas before but I've no idea why!

Comment (mjmottram) [inline src/plot/ROOTMultiPlot.cpp]: Wait, I remember the reason for this now (although maybe there isn't a use case): what if construct is called multiple times?  At least a clear is needed on the canvas here, right?

Comment (jackdunger) [inline src/plot/ROOTMultiPlot.cpp]: Ah ha I see. Well looking at it now i think fConstructed is a flag that stops that happening, and Construct() itself is private.. 

But at the moment, if you want to add an extra pdf after drawing it wont do that. I guess it needs a clear() method that wipes the canvas and sets fConstructed to false.

Comment (jackdunger) [inline src/plot/ROOTMultiPlot.cpp]: I tested this in the last week and it did what was expected for just adding a bunch of pdfs and then drawing. But probably wouldnt have if I added more stuff on top of that and drew again

**Comment by mjmottram** [comment]
👍 ? 👎 ?

**Comment by mjmottram** [comment]
@jackdunger I've added and tested function overloading both for different arguments and for default argument cases.

**Comment by mjmottram** [comment]
Didn't get around to working out how to return and plot ROOT objects from PdfConverter though.  Might revisit it at another point.

**Comment by jackdunger** [comment]
Hi @mjmottram, yes that one is harder. I did think of an intermediate idea though: we could have a save as root object method in pdf converter, with the saving done in c++

**Comment by jackdunger** [comment]
----------- testing if this message shows up -----------

**Comment by jackdunger** [comment]
@mjmottram, what do you think of this? https://github.com/pybind/pybind11

**Comment by mjmottram** [comment]
Looks nice.  You happy to enforce C++11?  And make me do more work?

**Comment by jackdunger** [comment]
header only, out of the box support for most stl containers

**Comment by jackdunger** [comment]
would have to think about it, I think the syntax is mostly the same. We wouldn't have to enforce C++11 everywhere. Would just really like these out of the box

**Comment by mjmottram** [comment]
OK, I'll have a look this week.

**Pull request closed without merging**